datafusion/optimizer/src/common_subexpr_eliminate.rs

```diff
@@ -714,14 +689,7 @@ impl<'n> TreeNodeVisitor<'n> for ExprIdentifierVisitor<'_> {
         self.id_array[down_index].0 = self.up_index;
         if !self.expr_mask.ignores(expr) {
             self.id_array[down_index].1.clone_from(&expr_id);
-
-            // TODO: can we capture the data type in the second traversal only for
-            //  replaced expressions?
-            let data_type = expr.get_type(&self.input_schema)?;
-            let (count, _) = self
-                .expr_stats
-                .entry(expr_id.clone())
-                .or_insert((0, data_type));
+            let count = self.expr_stats.entry(expr_id.clone()).or_insert(0);
```

Comment:
it would be great to avoid this clone (which I think is a String) by using a slightly different API. Maybe something like

```rust
let count = if let Some(count) = self.expr_stats.get(expr_id) {
  count
} else {
  self.expr_stats.insert(expr_id.clone(), 0);
  0
};
```

Which while less elegant and requires hashing twice doens't clone the string 🤔 

There may be a more clever way to avoid cloning the id

Also, I see the current code clones `expr_id` as well, so this change is no worse.

Reply:
The reason why I wouldn't change this in this PR is because my https://github.com/apache/datafusion/pull/10473 (https://github.com/apache/datafusion/pull/10473/files#diff-351499880963d6a383c92e156e75019cd9ce33107724a9635853d7d4cd1898d0R761) will completely eliminate string identifiers and removes these clones.